sparse/tests/test_coo.py

```diff
@@ -1527,3 +1527,35 @@ def test_invalid_iterable_error():
     with pytest.raises(ValueError):
         x = [((2.3, 4.5), 3.2)]
         COO.from_iter(x)
+
+
+def test_roll_coo():
+    # 1d
+    x = np.array([0, 1, 0, 2, 0, 3, 0, 0, 4, 0, 0, 0, 5, 0, 6])
+    xs = sparse.as_coo(x)
+    for sh in (0, 2, -2, 20, -20):
+        assert_eq(np.roll(x, sh), sparse.roll(xs, sh).todense())
```

Comment:
The `.todense()` can be removed.

Reply:
Interesting.... This causes the test to fail at `assert is_canonical(...)` what is this checking for? Do I need to resort the coordinates or something inside `sparse.roll(...)` so that I maintain lexographic order of the indices?

Reply:
`COO` arrays are meant to be immutable. This means you shouldn't construct and then change their `coords`/`data`.

What you should do instead is, make `coords`/`data` and pass that to the constructor.

`is_canonical` checks if the `coords` have any duplicates, and are sorted. This is needed by many operations internally. In your case, they're not sorted, because you're changing them directly in the object, instead of passing them to the constructor. I suggest you use `COO(..., has_duplicates=False)`